cocotb/clock.py

```diff
@@ -165,3 +168,134 @@ async def start(self, cycles=None, start_high=True):
 
     def __str__(self):
         return type(self).__qualname__ + "(%3.1f MHz)" % self.frequency
+
+
+class SimClock(BaseClock):
+    r"""High performance, configurable simulator clock driver.
+
+    This class is a wrapper for a C++ layer clock driver with methods for control.
+    Use :meth:`start` to start the clock driver and :meth:`stop` to stop it.
+    The :attr:`is_running` property returns whether the driver is active.
+
+    Example:
+
+    .. code-block:: python
+
+        # defaults to 50:50 duty cycle
+        c = SimClock(dut.clk, 100, units='ns')
+        c.start()
+
+    Args:
+        signal: The clock pin/signal to be driven.
+        period (int, float, 2-tuple): The clock period. Must convert to an even number of
+            timesteps. Provide a 2-tuple as (high_time, low_time) to set the duty cycle manually.
```

Comment:
```suggestion
            timesteps. Provide a 2-tuple as (start_value_time, end_value_time) to set the duty cycle manually.
```
(to account for `start_high`)?

Reply:
The tuple defines the clock period, length of high part followed by length of low part.
It's analogous to duty cycle, which defines the fraction of the period where the clock is high.

My understanding of `start_high` is that it sets which part of the clock period to start from, it doesn't change how the duty cycle is interpreted.

Reply:
One thing I do and I think is sort of common is: support only one way to construct an object, and support conveniences as named constructors using `@classmethod`.

Reply:
Which one should be the default?

Reply:
> My understanding of `start_high` is that it sets which part of the clock period to start from, it doesn't change how the duty cycle is interpreted or invert the clock.

Oh, interesting... it never occurred to me that it could be viewed like this, as that's not what I had in mind when I introduced `start_high` back then. I saw it as "invert the clock", but since there is no commonly agreed upon "normal"/"non-inverted" clock, I made it explicit in the name.

Maybe we should have a drawing in the docstring, like... (I need to change to a real computer for that)
```
         ┌─────────
─────────┘          
```

Reply:
I can use @themperek's `posedge_first` instead. I'll add clarification in the docstring as well.